jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -3414,18 +3414,30 @@ public String getSimpleName() {
 		}
 	}
 	String simpleName = baseType.getSimpleNameImpl();
+	String fullName = baseType.getName();
+	Class parent = baseType.getEnclosingClass();
 	if (simpleName == null) {
 		// either a base class, or anonymous class
-		if (baseType.getEnclosingObjectClass() != null) {
+		if (parent != null) {
 			simpleName = ""; //$NON-NLS-1$
 		} else {
 			// remove the package name
-			simpleName = baseType.getName();
-			int index = simpleName.lastIndexOf('.');
+			int index = fullName.lastIndexOf('.');
 			if (index != -1) {
-				simpleName = simpleName.substring(index+1);
+				simpleName = fullName.substring(index+1);
 			}
 		}
+	} else if (!fullName.endsWith(simpleName)) {
+		int index = fullName.lastIndexOf('.');
+		if (parent != null) {
+			String parentName = parent.getName();
+			if (fullName.startsWith(parentName) && (fullName.charAt(parentName.length()) == '$')) {
+				index = parentName.length();
+			}
+		}
+		if (index != -1) {
+			simpleName = fullName.substring(index+1);
+		}
```

Comment:
Please review JLS 13.1 for details on what constitutes a simple name: http://cr.openjdk.java.net/~iris/se/12/latestSpec/java-se-12-jls-diffs.pdf